content/blog/2020-04-dplyr-1-0-0-rowwise/index.Rmarkdown

```diff
@@ -0,0 +1,205 @@
+
+```{r}
+df <- tibble(
+  student_id = 1:4, 
+  test1 = 10:13, 
+  test2 = 20:23, 
+  test3 = 30:33, 
+  test4 = 40:43
+)
+df
+```
+
+It'd like to be able to compute the mean of the test scores for each student, but `mutate()` and `mean()` do what I want:
+
+```{r}
+df %>% mutate(avg = mean(c(test1, test2, test3, test4)))
+```
+
+The problem is that I'm getting a mean over the whole data frame, not for each student. I can resolve this problem a get a mean for each student by creating create a "row-wise" data frame with `rowwise()`:
+
+```{r}
+rf <- rowwise(df, student_id)
+```
+
+`rowwise()` doesn't need any additional arguments unless you have variables that identify the rows, like `student_id` here. Much like grouping variables, they will be automatically preserved when you `summarise()` the data.
+
+```{r}
+rf
+```
+
+`rf` _looks_ very similar to `df`, but _behaves_ very differently:
+
+```{r}
+rf %>% mutate(avg = mean(c(test1, test2, test3, test4)))
+```
+An additional advantage of `rowwise()` is that it's paired with `c_across()`, which works like `c()` but uses the same tidyselect syntax as `across()`. That makes it easy to operate on multiple variables:
+
+```{r}
+rf %>% mutate(avg = mean(c_across(starts_with("test"))))
+```
+### Other ways of achieving the same result
+
+Some summary functions have alternative ways to compute row-wise summaries that take advantage of built-in vectorisation. For example, if you wanted to compute the sum, you could use `+`:
+
+```{r}
+df %>% mutate(total = test1 + test2 + test3 + test4)
+```
+And you _could_ use the same basic idea to compute the mean:
+
+```{r}
+df %>% mutate(avg = (test1 + test2 + test3 + test4) / 4)
+```
+
+Another family of summary functions have "parallel" extensions where you can provide multiple variables in the arguments:
+
+```{r}
+df %>% mutate(
+  min = pmin(test1, test2, test3, test4), 
+  max = pmax(test1, test2, test3, test4), 
+  string = paste(test1, test2, test3, test4, sep = "-")
+)
+```
+
+Where these functions exist, they'll usually be faster than `rowwise()`. The advantage of `rowwise()` is that it works with any function, not just those that are already vectorised.
+
+## List-columns
+
+`rowwise()` is useful for computing simple summaries, but its real power comes when you use it with list-columns. Because lists can contain anything, you can use list-columns to keep related objects together, regardless of what type of thing they are. List-columns give you a convenient storage mechanism and `rowwise()` gives you a convenient computation mechanism.
+
+Let's make those ideas concrete by creating a data frame a list-column. A little later, we'll come back to how you might actually get a list-column in a more realistic situation. The following data frame uses list columns to store things that would otherwise be challenging:
```

Comment:
```suggestion
Let's make those ideas concrete by creating a data frame with a list-column. A little later, we'll come back to how you might actually get a list-column in a more realistic situation. The following data frame uses list columns to store things that would otherwise be challenging:
```